keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -92,51 +92,51 @@ type Coscheduling struct {
 
-**limition**: `QueueSort` is the core part of our design and only one `QueueSort` plugin is allowed in the scheduling framework. So our design only supports the case that `QueueSort` extension point isn't implemented in other plugins.
+**limitation**: `QueueSort` is the core part of our design and only one `QueueSort` plugin is allowed in the scheduling framework. So our design only supports the case that `QueueSort` extension point isn't implemented in other plugins.
 
 ```go
 func  Less(podA *PodInfo, podB *PodInfo) bool
 ```
-1. Trying to order by pod priority (i.e. .spec.priorityValue), pod with higher priority is scheduled ahead of other pod with lower priority. When priorities are the same, they are operated according to the following process. 
-   
-2. When podA and podB are both regularPods (we will check it by their labels), it follows the same logic of default in-tree [PrioritySort](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45) plugin.
-   
-3. When podA is regularPod, podB is pgPod, trying to order by podA's `Timestamp` (the time pod added to the scheduling queue) and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). Pod with earlier timestamp is scheduled ahead of other pod.   
-   
-4. When podA and podB are both pgPods. 
-   1. Trying to order by the `LastFailureTimestamp` of the podGroup. Pod with earlier timestamp is scheduled ahead of other pod.
-   
-   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, trying to order by the `UID` of `PodGroup`,`Pod` with lexicographically greater `UID` is scheduled ahead of other pod. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)
-
-**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.
+
+Firstly, we will inherit the default in-tree PrioritySort plugin so as to honor .spec.priority to ensure high-priority Pods are always sorted ahead of low-priority ones.
+
+Secondly, if two Pods hold the same priority, the sorting precedence is described as below:
+
+- If they are both regularPods (without particular PodGroup label), compare their `InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned ahead of the other.
+  
+- If one is regularPod and the other is pgPod, compare regularPod's `InitialAttemptTimestamp` with pgPod's `InitialTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
+  
+- If they are both pgPods:
+  - Compare their `InitialAttemptTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
```

Comment:
Have you thought of starvation?
If two pod groups have the same priority, the first one to enter the queue will always have higher precedence than the other, or even individual pods.

This is different than the original, where the scheduler tries to give the same scheduling opportunities to all pods. But, on the other hand, this might be something that we want for co-scheduling.

Ref https://github.com/kubernetes/kubernetes/issues/83834#issuecomment-587557978
cc @yqwang-ms

Reply:
Thanks for cc.

BTW, seems co-scheduling by itself just needs "Gang Queuing", i.e.:
> Pods belonging to the same PodGroup are queued back-to-back

Which is not necessary to be also FIFO. (And FIFO might NOT be something that we want for co-scheduling, it is more like that to implement co-scheduling, to be FIFO is a helper, so we coupled them)

Given current K8S has priority starvation issues mentioned in https://github.com/kubernetes/kubernetes/issues/86373.
E.g. an early large scale pod group, keeps in waiting, will also cause later small pod group to starve, even if the small pod group can fit into free resource, and from user view, they have the same priority, and just only a little different creation time.

So, we would better decouple FIFO with co-scheduling, such as, for sometime, we move the whole early group behind later group, and for other times, we move the whole back.

Reply:
> If two pod groups have the same priority, the first one to enter the queue will always have higher precedence than the other, or even individual pods.

Technically, due to the limitation of the current global one-queue mechanism, it's possible upon a large pod group as it's pretty challenging to accommodate all precedence into one queue.

> But, on the other hand, this might be something that we want for co-scheduling.

Comparing to the way that randomly sorts PodGroup Pods (where wait() is more unpredictable), this approach seems to have more chances to avoid starvation. However, as I mentioned above, technically it still has chances.

Reply:
> So, we would better decouple FIFO with co-scheduling, such as, for sometime, we move the whole early group behind later group, and for other times, we move the whole back.

Could you rephrase this? Did you mean couple or decouple? What would be "other times"?

Reply:
I mean we would better decouple them given current situation.
For example, a large early pod group A, and another small later pod group B.
For queuing order, could we make A has 70% chance ahead of B? (So B still has 30% chance ahead of A)? 
So in this way, we at least can mitigate or avoid the starvation for pods of the same priority. Otherwise, same priority pods also suffers from this proposal, right?

Reply:
I mean can we easily queue pods in a group as a gang/unit, instead of leveraging detailed tuned compare rules to make sure they are back-to-back (has side effects to be FIFO)?

Reply:
> I mean can we easily queue pods in a group as a gang/unit, instead of leveraging detailed tuned compare rules to make sure they are back-to-back (has side effects to be FIFO)?

I mentioned it in `Alternatives considered`. I think this requires major refactoring, which only supports Pods as scheduling unit today.  @yqwang-ms